opentelemetry-sdk/src/resource/builder.rs

```diff
@@ -0,0 +1,96 @@
+use opentelemetry::KeyValue;
+
+use super::{Resource, ResourceDetector};
+
+/// Builder to allow easy composition of a Resource
+#[derive(Debug, Default)]
+pub struct ResourceBuilder {
+    resource: Resource,
+}
+
+impl ResourceBuilder {
+    /// Create ResourceBuilder with an empty [Resource].
+    pub fn new_empty() -> Self {
+        ResourceBuilder {
+            resource: Resource::empty(),
+        }
+    }
+
+    /// Create ResourceBuilder with a default [Resource].
+    pub fn new_default() -> Self {
+        ResourceBuilder {
+            resource: Resource::default(),
+        }
+    }
+
+    /// Add a single [ResourceDetector] to your resource.
+    pub fn with_detector(self, detector: Box<dyn ResourceDetector>) -> Self {
+        self.with_detectors(vec![detector])
+    }
+
+    /// Add multiple [ResourceDetector] to your resource.
+    pub fn with_detectors(mut self, detectors: Vec<Box<dyn ResourceDetector>>) -> Self {
+        self.resource = self.resource.merge(&Resource::from_detectors(detectors));
+        self
+    }
+
+    /// Add a [KeyValue] to the resource.
+    pub fn with_key_value(self, kv: KeyValue) -> Self {
```

Comment:
interesting why this does not need `mut self`...

Reply:
It might be since the with_detectors consumes the builder and its not a `&mut` reference? That's just a guess I'm actually not sure about that 😅

Reply:
`with_key_value` doesn’t need `mut self` because it doesn’t directly modify the builder; it delegates to `with_key_values`, which takes mut self to perform the actual mutation.